File: system/platform-core/src/main/java/org/platformlambda/core/actuator/ActuatorServices.java
/*

    Copyright 2018-2023 Accenture Technology

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.

 */

package org.platformlambda.core.actuator;

import org.platformlambda.core.annotations.PreLoad;
import org.platformlambda.core.models.EventEnvelope;
import org.platformlambda.core.models.LambdaFunction;
import java.util.concurrent.atomic.AtomicBoolean;

import java.util.Map;

@PreLoad(route="actuator.services", instances=10)
public class ActuatorServices implements LambdaFunction {
    private static final String TYPE = "type";
    private static final String INFO = "info";
    private static final String ROUTES = "routes";
    private static final String LIB = "lib";
    private static final String ENV = "env";
    private static final String HEALTH = "health";
    private static final String HEALTH_STATUS = "health_status";
    private static final String SHUTDOWN = "shutdown";
    private static final String SUSPEND = "suspend";
    private static final String RESUME = "resume";
    private static final String LIVENESS_PROBE = "livenessprobe";
    private static final String USER = "user";

    private static final InfoService infoFunction = new InfoService();
    private static final HealthService healthFunction = new HealthService();
    private static final ShutdownService shutdownFunction = new ShutdownService();
    private static final SuspendResume suspendResume = new SuspendResume();
    private static final AtomicBoolean healthStatus = new AtomicBoolean(true);

    @Override
    public Object handleEvent(Map<String, String> headers, Object body, int instance) throws Exception {
        if (headers.containsKey(TYPE)) {
            String type = headers.get(TYPE);
            if (HEALTH_STATUS.equals(type) && body instanceof Boolean) {
                healthStatus.set((Boolean) body);
                return true;
            }
            if (LIVENESS_PROBE.equals(type)) {
                if (healthStatus.get()) {
                    return new EventEnvelope().setBody("OK").setHeader("content-type", "text/plain");
                } else {
                    return new EventEnvelope().setBody("Unhealthy. Please check '/health' endpoint.")
                                    .setStatus(400).setHeader("content-type", "text/plain");
                }
            }
            if (HEALTH.equals(type)) {
                return healthFunction.handleEvent(headers, body, instance);
            }
            if (INFO.equals(type) || LIB.equals(type) || ROUTES.equals(type) || ENV.equals(type)) {
                return infoFunction.handleEvent(headers, body, instance);
            }
            if (headers.containsKey(USER)) {
                if (SUSPEND.equals(type) || RESUME.equals(type)) {
                    return suspendResume.handleEvent(headers, body, instance);
                }
                if (SHUTDOWN.equals(type)) {
                    return shutdownFunction.handleEvent(headers, body, instance);
                }
            }
        }
        return false;
    }

}

File: system/platform-core/src/main/kotlin/org/platformlambda/core/actuator/ActuatorServices.kt
/*

    Copyright 2018-2024 Accenture Technology

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.

 */

package org.platformlambda.core.actuator

import org.platformlambda.core.annotations.PreLoad
import org.platformlambda.core.models.EventEnvelope
import org.platformlambda.core.models.KotlinLambdaFunction
import org.platformlambda.core.models.Kv
import org.platformlambda.core.system.EventEmitter
import org.platformlambda.core.system.ServiceDiscovery
import org.slf4j.LoggerFactory
import java.io.IOException
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.system.exitProcess

@PreLoad(route = "actuator.services", instances = 10)
class ActuatorServices : KotlinLambdaFunction<EventEnvelope, Any> {

    companion object {
        private val log = LoggerFactory.getLogger(ActuatorServices::class.java)
        private const val TYPE = "type"
        private const val INFO = "info"
        private const val ROUTES = "routes"
        private const val LIB = "lib"
        private const val ENV = "env"
        private const val HEALTH = "health"
        private const val HEALTH_STATUS = "health_status"
        private const val SHUTDOWN = "shutdown"
        private const val SUSPEND = "suspend"
        private const val RESUME = "resume"
        private const val LIVENESS_PROBE = "livenessprobe"
        private const val USER = "user"
        private const val WHEN = "when"
        private val infoFunction = InfoService()
        private val healthFunction = HealthService()
        private val healthStatus = AtomicBoolean(true)
    }

    @Throws(Exception::class)
    override suspend fun handleEvent(headers: Map<String, String>, input: EventEnvelope, instance: Int): Any {
        if (headers.containsKey(TYPE)) {
            val type = headers[TYPE]
            if (HEALTH_STATUS == type && input.body is Boolean) {
                healthStatus.set(input.body as Boolean)
                return true
            }
            if (LIVENESS_PROBE == type) {
                return if (healthStatus.get()) {
                    EventEnvelope().setBody("OK").setHeader("content-type", "text/plain")
                } else {
                    EventEnvelope().setBody("Unhealthy. Please check '/health' endpoint.")
                        .setStatus(400).setHeader("content-type", "text/plain")
                }
            }
            if (HEALTH == type) {
                return healthFunction.handleEvent(headers, input, instance)
            }
            if (INFO == type || LIB == type || ROUTES == type || ENV == type) {
                return infoFunction.handleEvent(headers, input, instance)
            }
            if (headers.containsKey(USER)) {
                if (SUSPEND == type || RESUME == type) {
                    if (headers.containsKey(USER) && headers.containsKey(TYPE) && headers.containsKey(WHEN)) {
                        try {
                            EventEmitter.getInstance().send(ServiceDiscovery.SERVICE_REGISTRY,
                                Kv(TYPE, headers[TYPE]), Kv(WHEN, headers[WHEN]), Kv(USER, headers[USER]))
                        } catch (e: IOException) {
                            log.error("Unable to perform {} - {}", type, e.message)
                        }
                    }
                    return false
                }
                if (SHUTDOWN == type && headers.containsKey(USER)) {
                    log.info("Shutdown requested by {}", headers[USER])
                    exitProcess(-2)
                }
            }
        }
        return false;
    }

}